### PR TITLE
Track Location Uptime

### PR DESCRIPTION
**Pull request by Ashesh3**
This pull request adds a new task to periodically check the status of asset locations and record their availability. It also includes changes to the API to support querying the availability records.

### Key File Changes
1. **Asset Serializer Modifications (`asset.py` in `serializers` and `viewsets`):**
   - Refactoring of `AssetAvailabilityRecord` to `AvailabilityRecord`.
   - Adjustment of serializer fields and methods to align with the new model structure.
   - Changes in the viewset filters to accommodate the updated model.

2. **Migration Script (`0408_availabilityrecord_delete_assetavailabilityrecord_and_more.py`):**
   - A comprehensive migration script to transition from `AssetAvailabilityRecord` to `AvailabilityRecord`.
   - Includes forward and backward migration logic to ensure database consistency.

3. **Model Changes (`asset.py` in `models`):**
   - Introduction of the `AvailabilityRecord` model.
   - Removal of `AssetAvailabilityRecord` and updates to related model fields.
   - Adjustments to model methods and properties to reflect new relationships and data structures.

4. **Task Updates (`asset_monitor.py` and `location_monitor.py` in `tasks`):**
   - Modification of task logic to use `AvailabilityRecord`.
   - Introduction of new periodic tasks for location monitoring (`location_monitor.py`).

5. **Minor Adjustments (`__init__.py` in `tasks`):**
   - Inclusion of the new location monitoring task in the task initialization.

### Summary
The PR aims to improve the tracking system for asset and location availability within the CARE project. The changes suggest a shift towards a more generic `AvailabilityRecord` model that can handle different types of assets and locations, enhancing the system's flexibility and scalability. The migration script ensures a smooth transition from the old model to the new system without data loss.

The modifications in serializers, viewsets, and models indicate a thorough update to accommodate these changes, ensuring that the system's API and internal workings remain consistent with the new data structure. Additionally, the updates to the periodic tasks hint at a broader application of these changes, likely impacting how asset and location statuses are monitored and recorded in real time. 

These changes reflect a significant improvement in the system's ability to track and report on the operational status of various assets and locations, which is crucial for efficient resource management and operational oversight in the context of the CARE project.

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

File: care/facility/api/viewsets/asset.py [hunk 1]
```diff
@@ -205,15 +208,28 @@ def retrieve(self, request, *args, **kwargs):
...
+    ListModelMixin, RetrieveModelMixin, GenericViewSet
+):
+    queryset = AvailabilityRecord.objects.all()
+    serializer_class = AvailabilityRecordSerializer
+
+    def get_queryset(self):
+        return self.queryset.filter(
+            object_external_id=self.kwargs["asset_location_external_id"],
+            content_type__model="assetlocation",
+        )
```
Comment (sainak) [inline care/facility/api/viewsets/asset.py:232]: needs access checks

you can use a common viewset for both

add permission classes

and for access control call get_facility_queryset and then try to resolve the object using the facility queryset

**Review by sainak — CHANGES_REQUESTED** [review]

File: care/facility/api/viewsets/asset.py (hunk 1 above)
Comment (Ashesh3) [reply care/facility/api/viewsets/asset.py]: Changes made and tested. Also added tests for the permission checks.

**Review by sainak — APPROVED** [review]
LGTM

**Review by rithviknishad — APPROVED** [review]

**Comment by Ashesh3** [comment]
@vigneshhari could you review this once?

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**